README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# httpproblems

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/httpproblems)](https://CRAN.R-project.org/package=httpproblems)
[![R-CMD-check](https://github.com/atheriel/httpproblems/workflows/R-CMD-check/badge.svg)](https://github.com/atheriel/httpproblems/actions)
<!-- badges: end -->

This R package is an answer to the question “how can I report errors
from my R API in a standard, consistent way?”

Specifically, it implements the “Problem Details” specification from
[RFC 7807](https://tools.ietf.org/html/rfc7807), which is an emerging
standard for errors supported by many other languages and tools.

Although **httpproblems** originated for use with
[Plumber](https://www.rplumber.io) (see below), it has no dependencies
of its own and can be used with any R web framework.

## Installation

You can install the latest release of **httpproblems** from
[CRAN](https://CRAN.R-project.org):

``` r
install.packages("httpproblems")
```

or the development version from GitHub with

``` r
# install.packages("remotes")
remotes::install_github("atheriel/httpproblems")
```

## Usage

**httpproblems** provides helpers to create a few common HTTP errors:

  - `bad_request()`
  - `not_found()`
  - `unauthorized()`
  - `forbidden()`
  - `conflict()`
  - `internal_server_error()`

as well as a generic `http_problem()` that can be used with any valid
status code.

These functions return a normal list with elements [in line with the
standard](https://tools.ietf.org/html/rfc7807#section-3.1):

``` r
library(httpproblems)

body <- bad_request("Parameter 'id' must be a number.")
str(body)
#> List of 4
#>  $ type  : chr "https://tools.ietf.org/html/rfc7231#section-6.5.1"
#>  $ title : chr "Bad Request"
#>  $ status: int 400
#>  $ detail: chr "Parameter 'id' must be a number."
#>  - attr(*, "class")= chr [1:2] "http_problem" "list"
```

This object could be returned by an endpoint as normal after
serialization to e.g. JSON.

In addition to these basic methods, **httpproblems** also has a set of
`stop_for_http_problem()` functions:

  - `stop_for_bad_request()`
  - `stop_for_not_found()`
  - `stop_for_unauthorized()`
  - `stop_for_forbidden()`
  - `stop_for_conflict()`
  - `stop_for_internal_server_error()`

These leverage R’s extensible condition system to issue custom errors
for these problems:

``` r
stop_for_bad_request("Parameter 'id' must be a number.")
#> Error: Bad Request (HTTP 400). Parameter 'id' must be a number.

# What does the error actually contain?
tryCatch(
  stop_for_bad_request("Parameter 'id' must be a number."),
  error = function(e) {
    str(e)
  }
)
#> List of 3
#>  $ message: chr "Bad Request (HTTP 400). Parameter 'id' must be a number."
#>  $ status : int 400
#>  $ body   :List of 4
#>   ..$ type  : chr "https://tools.ietf.org/html/rfc7231#section-6.5.1"
#>   ..$ title : chr "Bad Request"
#>   ..$ status: int 400
#>   ..$ detail: chr "Parameter 'id' must be a number."
#>   ..- attr(*, "class")= chr [1:2] "http_problem" "list"
#>  - attr(*, "class")= chr [1:4] "http_problem_error" "http_error" "error" "condition"
```

Typically, web frameworks allow you to control how uncaught errors are
reported – which can be combined with these helpers to ensure that all
API errors have a consistent format for your users.

### Use with Plumber

Plumber allows users to control how errors are handled (via
`pr_set_error()`), which we can employ to produce a Problem Details
structure for both expected *and* unexpected errors:

``` r
library(plumber)

pr() %>%
  pr_get("/bad", function(req, res) {
    stop_for_bad_request("The 'id' parameter must be a number.")
  }) %>%
  pr_get("/bug", function(req, res) {
    stop("Another R error.")
  }) %>%
  pr_set_error(function(req, res, err) {
    # Force "unboxed" JSON and the Content-Type from RFC 7807.
    res$serializer <- serializer_unboxed_json(
      type = "application/problem+json"
    )
    # If we have an http_problem_error, use its status and body
    # fields. Otherwise, issue a 500 Internal Server Error.
    if (inherits(err, "http_problem_error")) {
      res$status <- err$status
      return(err$body)
    }
    if (isTRUE(req$pr$getDebug())) {
      internal_server_error(detail = err$message)
    } else {
      internal_server_error()
    }
  }) %>%
  pr_run(port = 4444)
```

Users interacting with this API will see the following:

``` shell
$ curl -vs localhost:4444/bad
> GET /bad HTTP/1.1
> Host: localhost:4444
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Date: Tue, 09 Feb 2021 15:19:33 GMT
< Content-Type: application/problem+json
< Content-Length: 160
< 
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "Bad Request",
  "status": 400,
  "detail": "The 'id' parameter must be a number."
}
$ curl -vs localhost:4444/bug
> GET /bug HTTP/1.1
> Host: localhost:4444
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 09 Feb 2021 15:19:59 GMT
< Content-Type: application/problem+json
< Content-Length: 150
< 
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.6.1",
  "title": "Internal Server Error",
  "status": 500,
  "detail": "Another R error."
}
```

You may also wish to use `not_found()` for Plumber’s 404 handler via
`pr_set_404()`.

## License

Copyright 2021 Aaron Jacobs

Licensed under the Apache License, Version 2.0 (the “License”); you may
not use this file except in compliance with the License. You may obtain
a copy of the License at

<https://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


NEWS.md
# httpproblems 1.0.1.9000

* Error messages for invalid inputs have been improved.

# httpproblems 1.0.1

* Removes the `LazyData` field from the `DESCRIPTION` file, which was generating
  a NOTE in CRAN's automated checking.

# httpproblems 1.0.0

* Initial release. **httpproblems** provides tools for emitting the "Problem
  Details" structure defined in [RFC 7807](https://tools.ietf.org/html/rfc7807).
  It is intended to help R users report errors from HTTP APIs in a consistent,
  standard way.
  
* Provides `http_problem()` for creating the Problem Details structure, plus
  several helper functions for HTTP 400 Bad Request, 404 Not Found, 401
  Unauthorized, 403 Forbidden, 409 Conflict, and 500 Internal Server Error.

* Provides `stop_for_*()` variants that signal a custom condition for these
  problems.

* Provides `http_problem_types()` for listing built-in type and title entries.


tests/testthat.R
library(testthat)
library(httpproblems)

test_check("httpproblems")


tests/tinytest.R
library(tinytest)

test_package("httpproblems")


tests/testthat/test-problems.R
expect_http_problem <- function(x, status = 500L) {
  testthat::expect_s3_class(x, "http_problem")
  testthat::expect_true(is.character(x$type))
  testthat::expect_true(is.character(x$title))
  testthat::expect_equal(x$status, status)
}

expect_http_error <- function(expr, status = 500L) {
  x <- tryCatch(expr, error = function(e) e)
  testthat::expect_s3_class(x, "error")
  testthat::expect_s3_class(x, "http_problem_error")
  expect_http_problem(x$body, status)
  testthat::expect_equal(x$status, status)
  testthat::expect_match(x$message, sprintf("HTTP %d", status))
  testthat::expect_match(x$message, x$body$title)
  if (!is.null(x$body$detail)) {
    testthat::expect_match(x$message, x$body$detail)
  }
  x
}

testthat::test_that("http_problem structures work as expected", {
  out <- http_problem()
  expect_http_problem(out)
  testthat::expect_true(is.null(out$detail))
  testthat::expect_true(is.null(out$instance))
})

testthat::test_that("http_problems with optional fields work as expected", {
  out <- http_problem(details = "Unknown", instance = "/widgets/101")
  testthat::expect_true(is.character(out$detail))
  testthat::expect_true(is.character(out$instance))
})

testthat::test_that("http_problem structures can be extended", {
  out <- http_problem(random_code = 46L)
  expect_http_problem(out)
  testthat::expect_equal(out$random_code, 46L)
})

testthat::test_that("http_problem conditions work as expected", {
  out <- expect_http_error(stop_for_http_problem())
  testthat::expect_true(is.null(out$body$detail))
  testthat::expect_true(is.null(out$body$instance))
  out <- expect_http_error(
    stop_for_http_problem(details = "Unknown", instance = "/widgets/101")
  )
  testthat::expect_true(is.character(out$body$detail))
  testthat::expect_true(is.character(out$body$instance))
})

testthat::test_that("http_problem helpers work as expected", {
  expect_http_problem(bad_request(), 400L)
  expect_http_problem(unauthorized(), 401L)
  expect_http_problem(forbidden(), 403L)
  expect_http_problem(not_found(), 404L)
  expect_http_problem(conflict(), 409L)
  expect_http_problem(internal_server_error(), 500L)

  expect_http_error(stop_for_bad_request(), 400L)
  expect_http_error(stop_for_unauthorized(), 401L)
  expect_http_error(stop_for_forbidden(), 403L)
  expect_http_error(stop_for_not_found(), 404L)
  expect_http_error(stop_for_conflict(), 409L)
  expect_http_error(stop_for_internal_server_error(), 500L)
})


R/package.R
# Copyright 2021 Aaron Jacobs
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#' Describe a Problem with an HTTP Request
#'
#' @description
#'
#' `http_problem()` creates the "Problem Details" structure defined in
#' [RFC 7807](https://tools.ietf.org/html/rfc7807), used for reporting errors
#' from HTTP APIs in a standard way.
#'
#' There are also helper methods for the most common HTTP problems:
#' HTTP 400 Bad Request, 404 Not Found, 401 Unauthorized, 403 Forbidden, 409
#' Conflict, and 500 Internal Server Error.
#'
#' @param detail A human-readable string giving more detail about the error,
#'   if possible.
#' @param status The HTTP status code appropriate for the response.
#' @param type A URL pointing to human-readable documentation for this type of
#'   problem. When `NULL`, the type is generated based on the status code; see
#'   [http_problem_types()] for a list of the defaults.
#' @param title A ["short, human-readable summary of the problem type"](https://tools.ietf.org/html/rfc7807#section-3.1).
#'   When `NULL`, the title is generated based on the status code; see
#'   [http_problem_types()] for a list of the defaults.
#' @param instance A URL that identifies the specific occurrence of the
#'   problem, if possible. When `NULL` this field is simply excluded.
#' @param ... Additional fields added to the problem as [Extension Members](https://tools.ietf.org/html/rfc7807#section-3.2).
#'
#' @return An object of class `"http_problem"`, which has fields corresponding
#'   to [an RFC 7807 Problem Details structure](https://tools.ietf.org/html/rfc7807#section-3.1).
#'
#' @examples
#' body <- bad_request("Parameter 'id' must be a number.")
#' str(body)
#' @seealso [stop_for_http_problem] for issuing R errors with these structures.
#' @export
http_problem <- function(detail = NULL, status = 500L, type = NULL,
                         title = NULL, instance = NULL, ...) {
  if (!is.null(detail) && !is.character(detail)) {
    stop(
      sprintf("'detail' must be a string or omitted, not '%s'.", detail),
      call. = FALSE
    )
  }
  if (!is.null(instance) && !is.character(instance)) {
    stop(
      sprintf("'instance' must be a string or omitted, not '%s'.", instance),
      call. = FALSE
    )
  }
  if (!is.numeric(status)) {
    stop("'status' must be an HTTP status code, not '%s'.", call. = FALSE)
  }
  # Ignore vector arguments.
  status <- as.integer(status[1L])
  status_index <- which(status == http_problem_codes$code)
  if (length(status_index) == 0) {
    # NOTE: It's possible that we could allow arbitrary codes and just set
    # "about:blank" as the type (which would be in line with the RFC), but it
    # seems more likely that an unknown code is a programmer error.
    stop(sprintf("Unsupported HTTP status code for reporting problems: '%s'.", status))
  }
  # When not present, use standard title/type fields based on the status code.
  if (is.null(title)) {
    title <- http_problem_codes$reason[status_index]
  }
  if (is.null(type)) {
    type <- http_problem_codes$url[status_index]
  }
  body <- list(type = type[1L], title = title[1L], status = status, ...)
  # Only included optional fields if they are non-NULL.
  body$detail <- detail[1L]
  body$instance <- instance[1L]
  structure(body, class = c("http_problem", "list"))
}

#' @rdname http_problem
#' @export
bad_request <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 400L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname http_problem
#' @export
unauthorized <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 401L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname http_problem
#' @export
forbidden <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 403L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname http_problem
#' @export
not_found <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 404L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname http_problem
#' @export
conflict <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 409L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname http_problem
#' @export
internal_server_error <- function(detail = NULL, instance = NULL, ...) {
  http_problem(
    detail = detail, status = 500L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' Signal an Error Caused by an HTTP Problem
#'
#' @description
#'
#' The various `stop_for_*()` functions leverage R's condition system to signal
#' an error with a custom type embedding the "Problem Details" structure
#' defined in [RFC 7807](https://tools.ietf.org/html/rfc7807).
#'
#' They can be used for reporting errors from HTTP APIs in a standard way.
#'
#' There are also helper methods for the most common HTTP problems:
#' HTTP 400 Bad Request, 404 Not Found, 401 Unauthorized, 403 Forbidden, 409
#' Conflict, and 500 Internal Server Error.
#'
#' @return These functions call `stop()` with a custom [condition] (with class
#'   `"http_problem_error"`), so they do not return a value.
#'
#' @examples
#' tryCatch(
#'   stop_for_bad_request("Parameter 'id' must be a number."),
#'   error = function(e) {
#'     str(e)
#'   }
#' )
#' @inheritParams http_problem
#' @seealso [http_problem] for creating the structure directly.
#' @export
stop_for_http_problem <- function(detail = NULL, status = 500L, type = NULL,
                                  title = NULL, instance = NULL, ...) {
  problem <- http_problem(
    detail = detail, status = status, type = NULL, title = NULL,
    instance = instance, ...
  )
  if (is.null(problem$detail)) {
    message <- sprintf("%s (HTTP %d).", problem$title, problem$status)
  } else {
    message <- sprintf(
      "%s (HTTP %d). %s", problem$title, problem$status, problem$detail
    )
  }
  cond <- structure(
    list(message = message, status = problem$status, body = problem),
    class = c("http_problem_error", "http_error", "error", "condition")
  )
  # NOTE: stop() will also call signalCondition().
  stop(cond)
}

#' @rdname stop_for_http_problem
#' @export
stop_for_bad_request <- function(detail = NULL, instance = NULL, ...) {
  stop_for_http_problem(
    detail = detail, status = 400L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname stop_for_http_problem
#' @export
stop_for_unauthorized <- function(detail = NULL, instance = NULL, ...) {
  stop_for_http_problem(
    detail = detail, status = 401L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname stop_for_http_problem
#' @export
stop_for_forbidden <- function(detail = NULL, instance = NULL, ...) {
  stop_for_http_problem(
    detail = detail, status = 403L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname stop_for_http_problem
#' @export
stop_for_not_found <- function(detail = NULL, instance = NULL, ...) {
  stop_for_http_problem(
    detail = detail, status = 404L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname stop_for_http_problem
#' @export
stop_for_conflict <- function(detail = NULL, instance = NULL, ...) {
  stop_for_http_problem(
    detail = detail, status = 409L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

#' @rdname stop_for_http_problem
#' @export
stop_for_internal_server_error <- function(detail = NULL, instance = NULL,
                                           ...) {
  stop_for_http_problem(
    detail = detail, status = 500L, type = NULL, title = NULL,
    instance = instance, ...
  )
}

# Adapted from the table at https://tools.ietf.org/html/rfc7231#section-6.1
http_problem_codes <- data.frame(
  code = c(
    400L,
    401L,
    402L,
    403L,
    404L,
    405L,
    406L,
    407L,
    408L,
    409L,
    410L,
    411L,
    412L,
    413L,
    414L,
    415L,
    416L,
    417L,
    426L,
    500L,
    501L,
    502L,
    503L,
    504L,
    505L
  ),
  reason = c(
    "Bad Request",
    "Unauthorized",
    "Payment Required",
    "Forbidden",
    "Not Found",
    "Method Not Allowed",
    "Not Acceptable",
    "Proxy Authentication Required",
    "Request Timeout",
    "Conflict",
    "Gone",
    "Length Required",
    "Precondition Failed",
    "Payload Too Large",
    "URI Too Long",
    "Unsupported Media Type",
    "Range Not Satisfiable",
    "Expectation Failed",
    "Upgrade Required",
    "Internal Server Error",
    "Not Implemented",
    "Bad Gateway",
    "Service Unavailable",
    "Gateway Timeout",
    "HTTP Version Not Supported"
  ),
  url = c(
    "https://tools.ietf.org/html/rfc7231#section-6.5.1",
    "https://tools.ietf.org/html/rfc7235#section-3.1",
    "https://tools.ietf.org/html/rfc7231#section-6.5.2",
    "https://tools.ietf.org/html/rfc7231#section-6.5.3",
    "https://tools.ietf.org/html/rfc7231#section-6.5.4",
    "https://tools.ietf.org/html/rfc7231#section-6.5.5",
    "https://tools.ietf.org/html/rfc7231#section-6.5.6",
    "https://tools.ietf.org/html/rfc7235#section-3.2",
    "https://tools.ietf.org/html/rfc7231#section-6.5.7",
    "https://tools.ietf.org/html/rfc7231#section-6.5.8",
    "https://tools.ietf.org/html/rfc7231#section-6.5.9",
    "https://tools.ietf.org/html/rfc7231#section-6.5.10",
    "https://tools.ietf.org/html/rfc7232#section-4.2",
    "https://tools.ietf.org/html/rfc7231#section-6.5.11",
    "https://tools.ietf.org/html/rfc7231#section-6.5.12",
    "https://tools.ietf.org/html/rfc7231#section-6.5.13",
    "https://tools.ietf.org/html/rfc7233#section-4.4",
    "https://tools.ietf.org/html/rfc7231#section-6.5.14",
    "https://tools.ietf.org/html/rfc7231#section-6.5.15",
    "https://tools.ietf.org/html/rfc7231#section-6.6.1",
    "https://tools.ietf.org/html/rfc7231#section-6.6.2",
    "https://tools.ietf.org/html/rfc7231#section-6.6.3",
    "https://tools.ietf.org/html/rfc7231#section-6.6.4",
    "https://tools.ietf.org/html/rfc7231#section-6.6.5",
    "https://tools.ietf.org/html/rfc7231#section-6.6.6"
  ),
  stringsAsFactors = FALSE
)

#' List Built-In Problem Types
#'
#' Many APIs will not need to define custom problem "types", since HTTP status
#' codes [are usually illustrative enough](https://tools.ietf.org/html/rfc7807#section-4).
#' This function lists the default type and title information for a given
#' status code.
#'
#' @return A data frame of HTTP status codes and their default title & type.
#' @export
http_problem_types <- function() {
  # Reorder/rename the RFC table columns to make them look as they would in the
  # http_problem() output.
  out <- http_problem_codes[, c("code", "url", "reason")]
  names(out) <- c("status", "type", "title")
  out
}
